lib/adobexd.js

```diff
@@ -43,22 +43,42 @@ function filterReferences(references, filter) {
   return { key: filter, elements: filteredElements };
 }
 
+function findArtboardByID(artboards, id) {
+  const board = artboards.find(b => b.id === id.replace('artboard-', '')); 
+
+  return board.artboard.children[0];
+}
+
 async function parseXDFile(xdFilePath, outputDir) {
-  
   await unzipFile(xdFilePath, outputDir);
 
   const MANIFEST_FILEPATH = path.join(outputDir, 'manifest');
   const xdManifest = JSON.parse(await readFile(MANIFEST_FILEPATH));
 
-  const artboards = await getArtboardFiles(outputDir);
-
   const tokenArtboards = ['primitives/color', 'primitives/blur', 'primitives/border', 'primitives/gradient', 'textstyle'];
   const tokenReferences = getReferencesFromManifest(xdManifest);
   const groupedTokenReferences = tokenArtboards.map(key => filterReferences(tokenReferences, key));
 
+  const artboards = await getArtboardFiles(outputDir);
+
+  const matchedTokens = groupedTokenReferences.map(tokenGroup => tokenGroup.elements.map(ref => {
+    const { name } = ref;
+    const value = findArtboardByID(artboards, ref.pathToToken); 
+
+    return { name, value };
+  }));
+
+  let [colors, blurs, borders, gradients, textStyles] = matchedTokens;
+
+  colors = colors.map(color => ({ ...color, value: color.value.group.children[0].style.fill.color.value }));
+  blurs = blurs.map(blur => ({ ...blur, value: blur.value })); 
+  borders = borders.map(border => ({ ...border, value: border.value.style.stroke }));
+  gradients = gradients.map(gradient => ({ ...gradient, value: gradient.value.style.fill.gradient }));
+  textStyles = textStyles.map(textStyle => ({ ...textStyle, value: textStyle.value.meta.ux.rangedStyles[0] }));
+
   await removeDirectory(outputDir);
 
-  return { xdManifest, artboards, groupedTokenReferences };
+  return { colors, blurs, borders, gradients, textStyles };
```

Comment:
Why mutate these arrays? You can just return them:

```ts
const [colorTokens] = matchedTokens;

return {
  colors: colorTokens.map(color => ({ …color, value: … })),
}
```

Reply:
Also wondering if this mapping is something that should happen in the mappers instead of here.
In the parser, we basically just want: 

1. Get the filepath of XD file
2. Extract it to readable JSON files
3. Walk the JSON files and their references
4. Find anything that's a valid token (color, etc)
5. Get the object for that token in Adobe XD's JSON
6. Return that object under a property in the object returned from the parser

Parser = getting tokens
Mappers = from each specific design tool, map their design data to valid Hubble tokens